### PR TITLE
Start cleaning up tool resolver implementations.

### PR DESCRIPTION
**Pull request by grp**

### CODE REVIEW

File: Libraries/pbxbuild/Headers/pbxbuild/Tool/Input.h [hunk 1]
```diff
@@ -29,51 +29,63 @@ class Environment;
...
+     * The resolved absolute path to the file.
      */
-    Target::BuildRules::BuildRule::shared_ptr const &buildRule() const
-    { return _buildRule; }
+    std::string const &path() const
+    { return _path; }
 
     /*
-     * The type of this file.
+     * The type of this file. Optional.
```
Comment (Ktwu) [inline Libraries/pbxbuild/Headers/pbxbuild/Tool/Input.h]: Optional?  Then why doesn't this return an optional type? :)

File: Libraries/pbxbuild/Headers/pbxbuild/Tool/Input.h [hunk 2]
```diff
@@ -29,51 +29,63 @@ class Environment;
...
-     * The resolved absolute path to the file.
+     * The build rule saying *how* to build this file. Optional.
      */
-    std::string const &path() const
-    { return _path; }
+    Target::BuildRules::BuildRule::shared_ptr const &buildRule() const
+    { return _buildRule; }
+
+    /*
+     * A disambiguation identifier for files that have multiple with the
```
Comment (Ktwu) [inline Libraries/pbxbuild/Headers/pbxbuild/Tool/Input.h]: "that have multiple with the same base name"

Confusing.  You probably ought to reword this to something like "A disambiguation identifier for different files with the same base name..."

File: Libraries/pbxbuild/Headers/pbxbuild/Tool/Input.h [hunk 3]
```diff
@@ -29,51 +29,63 @@ class Environment;
  */
```
Comment (Ktwu) [inline Libraries/pbxbuild/Headers/pbxbuild/Tool/Input.h]: Remove reference to project objects from tool input: LGTM

File: Libraries/pbxbuild/Sources/Phase/Context.cpp [hunk 4]
```diff
@@ -263,7 +263,7 @@ Group(std::vector<Tool::Input> const &files)
          * the ungrouped  inputs so they don't get added again to the result.
          */
         ungrouped.erase(std::remove_if(ungrouped.begin(), ungrouped.end(), [&](Tool::Input const &ungroupedFile) {
-            if (ungroupedFile.fileType()->identifier() == "text.plist.strings") {
+            if (ungroupedFile.fileType() != nullptr && ungroupedFile.fileType()->identifier() == "text.plist.strings") {
```
Comment (Ktwu) [inline Libraries/pbxbuild/Sources/Phase/Context.cpp]: Once again, this seems like it ought to be an optional.

File: Libraries/pbxbuild/Sources/Phase/Context.cpp [hunk 5]
```diff
@@ -323,24 +324,28 @@ resolveBuildFiles(
 
             if (buildRule != nullptr) {
                 if (pbxspec::PBX::Tool::shared_ptr const &tool = buildRule->tool()) {
-                    // Some tools additionally limit their file types beyond what their build rule allows.
-                    // For example, the default compiler limits itself to just source files, despite its
-                    // default build rule specifying that it accepts all C-family inputs, including headers.
+                    /*
+                     * Some tools additionally limit their file types beyond what their build rule allows.
+                     * For example, the default compiler limits itself to just source files, despite its
+                     * default build rule specifying that it accepts all C-family inputs, including headers.
```
Comment (Ktwu) [inline Libraries/pbxbuild/Sources/Phase/Context.cpp]: Why is this burden on the tool?  Why wouldn't the default compiler's default build rule exclude headers?

File: Libraries/pbxbuild/Sources/Phase/FrameworksResolver.cpp [hunk 6]
```diff
@@ -84,21 +84,23 @@ resolve(Phase::Environment const &phaseEnvironment, Phase::Context *phaseContext
...
             if (it != phaseContext->toolContext().variantArchitectureInvocations().end()) {
                 std::vector<Tool::Invocation> const &sourceInvocations = it->second;
                 for (Tool::Invocation const &invocation : sourceInvocations) {
                     for (std::string const &output : invocation.outputs()) {
                         // TODO(grp): Is this the right set of source outputs to link?
+                        // TODO(grp): Use the object file file type and include in input.
                         if (FSUtil::GetFileExtension(output) == "o") {
-                            sourceOutputs.push_back(output);
+                            Tool::Input outputInput = Tool::Input(output, nullptr);
+                            sourceOutputs.push_back(outputInput);
```
Comment (Ktwu) [inline Libraries/pbxbuild/Sources/Phase/FrameworksResolver.cpp]: sourceOutputs.push_Back(Tool::Input(output, nullptr));

File: Libraries/pbxbuild/Headers/pbxbuild/Tool/CopyResolver.h [hunk 7]
```diff
@@ -34,13 +34,6 @@ class CopyResolver {
         std::string const &outputDirectory,
```
Comment (Ktwu) [inline Libraries/pbxbuild/Headers/pbxbuild/Tool/CopyResolver.h]: Always pass in inputs as tool inputs, rather than path strings: LGTM since I commented with nits.

File: Libraries/pbxbuild/Headers/pbxbuild/Phase/Context.h [hunk 8]
```diff
@@ -87,14 +87,14 @@ class Context {
     /*
      * Groups the files according to the tools used to build them.
      */
-    static std::vector<std::vector<Phase::File>> Group(std::vector<Phase::File> const &files);
+    static std::vector<std::vector<Tool::Input>> Group(std::vector<Tool::Input> const &files);
```
Comment (grp) [inline Libraries/pbxbuild/Headers/pbxbuild/Phase/Context.h]: Yeah, they're a wrapper around a file path with extra info.

File: Libraries/pbxbuild/Sources/Phase/Context.cpp (hunk 4 above)
Comment (grp) [inline Libraries/pbxbuild/Sources/Phase/Context.cpp]: `pbxspec::PBX::Specification` has to be a pointer since it's `virtual`. And `optional<SomePointer>` is confusing because it could be `nullopt`, `nullptr`, or a value.

Two options: make a non-nullable pointer type and wrap that in `optional`, or check `nullptr` like this does. Neither are great. (Rust gets this right, too...)

File: Libraries/pbxbuild/Sources/Phase/Context.cpp (hunk 5 above)
Comment (grp) [inline Libraries/pbxbuild/Sources/Phase/Context.cpp]: The issue is that it takes `sourcecode.cpp`, and `sourcecode.cpp.header` inherits from `sourcecode.cpp`. I tried not checking the inheritance hierarchy, but that failed for `text.plist.xml` inherits from `text.plist` and needs to be an input to the plist copy tool...

File: Libraries/pbxbuild/Headers/pbxbuild/Tool/Input.h (hunk 2 above)
Comment (grp) [inline Libraries/pbxbuild/Headers/pbxbuild/Tool/Input.h]: 👍

File: Libraries/pbxbuild/Headers/pbxbuild/Tool/AssetCatalogResolver.h [hunk 9]
```diff
@@ -10,14 +10,20 @@
 #ifndef __pbxbuild_Tool_AssetCatalogResolver_h
```
Comment (Ktwu) [inline Libraries/pbxbuild/Headers/pbxbuild/Tool/AssetCatalogResolver.h]: Start cleaning up tool resolver implementations: LGTM

File: Libraries/pbxbuild/Headers/pbxbuild/Tool/Context.h [hunk 10]
```diff
@@ -29,24 +29,27 @@ namespace Tool {
 
 class Context {
 private:
-    xcsdk::SDK::Target::shared_ptr _sdk;
+    xcsdk::SDK::Target::shared_ptr   _sdk;
     std::vector<xcsdk::SDK::Toolchain::shared_ptr> _toolchains;
```
Comment (Ktwu) [inline Libraries/pbxbuild/Headers/pbxbuild/Tool/Context.h]: Ugggh the spacing here for _toolchains is unsatisfying.

File: Libraries/pbxbuild/CMakeLists.txt [hunk 11]
```diff
@@ -13,6 +13,7 @@ add_library(pbxbuild SHARED
             Sources/DerivedDataHash.cpp
```
Comment (Ktwu) [inline Libraries/pbxbuild/CMakeLists.txt]: > Auxiliary files were attached to invocations but had no relation to
the containing invocation.

I see, so auxiliary files are basically used by Ninja, not the tools that Ninja invokes.

I actually thought it was nice that invocations contained their auxiliary files, since in a way such files a dependency for the invocation.

Oh well.  LGTM

File: Libraries/pbxbuild/Headers/pbxbuild/Phase/File.h [hunk 12]
```diff
@@ -29,73 +26,16 @@ class Environment;
...
-    /*
-     * Resolves build files into `Phase::File`s, using the phase environment. Note that due to
-     * certain types of build files being more complex than a single `Phase::File`, the number
-     * of `Phase::File`s returned may be different than the number of build files passed in.
+     * Resolves build files into `Tool::Input`s, using the phase environment. Note that due to
+     * certain types of build files being more complex than a single `Tool::Input`, the number
+     * of `Tool::Input`s returned may be different than the number of build files passed in.
      */
-    static std::vector<Phase::File>
+    static std::vector<Tool::Input>
```
Comment (Ktwu) [inline Libraries/pbxbuild/Headers/pbxbuild/Phase/File.h]: Ok.

File: Libraries/pbxbuild/Sources/Phase/Context.cpp (hunk 5 above)
Comment (Ktwu) [inline Libraries/pbxbuild/Sources/Phase/Context.cpp]: > sourcecode.cpp.header inherits from sourcecode.cpp
:( ok

File: Libraries/pbxbuild/CMakeLists.txt (hunk 11 above)
Comment (grp) [inline Libraries/pbxbuild/CMakeLists.txt]: Yeah, that was the original idea. I'd tried using that to put them in the dependency graph, but projects expect all of them to be written first and then used in custom scripts, etc. So they're already all being written before any invocations run.

In a some cases (scripts, linker file lists, Swift input) they were attached to the invocations that needed them. In a few cases (PCH info) they were totally independent. In others (header maps, module maps) they were dependencies but attached to the wrong invocation because the right one wasn't available.

Fixing the last set to work as dependencies was going to be pretty complicated especially for ones like module maps that are shared between multiple invocations, and not too useful due to the script input issue, so I sorta gave up on the idea.

**Comment by Ktwu** [comment]
I don't think I had any major recommendations so overall LGTM.
